### PR TITLE
update image segementation

### PR DESCRIPTION
**Pull request by MorganeLu**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] [local build the book](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-book)
- [ ] review the artifacts from local build
- [ ] [local build the slides](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-slides-optional) if appliable
- [ ] Unit test if appliable

# Checklist:

- [x] My code follows the [style guidelines](https://github.com/open-academy/machine-learning/blob/main/STYLE_GUIDE.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have applied tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md [hunk 1]
```diff
@@ -0,0 +1,973 @@
...
+import matplotlib.pyplot as plt
+```
+
+### Prepare the Oxford-IIIT Pets dataset
+
+```{code-cell}
+dataset, info = tfds.load('oxford_iiit_pet:3.*.*', with_info=True)
+```
+
+In addition, the image color values are normalized to the range. Finally, as mentioned above the pixels in the segmentation mask are labeled either {1, 2, 3}. For the sake of convenience, subtract 1 from the segmentation mask, resulting in labels that are : {0, 1, 2}.[0, 1]
```
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:50]: [0, 1] ?

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+This tutorial uses the Oxford-IIIT Pet Dataset. The dataset consists of images of 37 pet breeds, with 200 images per breed (~100 each in the training and test splits). Each image includes the corresponding labels, and pixel-wise masks. The masks are class-labels for each pixel. Each pixel is given one of three categories:
+
+- Class 1: Pixel belonging to the pet.
+- Class 2: Pixel bordering the pet.
+- Class 3: None of the above/a surrounding pixel.
+
+And it can be downloaded from [The Oxford-IIIT Pet Dataset](https://academictorrents.com/details/b18bbd9ba03d50b0f7f479acc9f4228a408cecc1).
+
+```{code-cell}
+pip install git+https://github.com/tensorflow/examples.git
```
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:32]: ?

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+  x = last(x)
+
+  return tf.keras.Model(inputs=inputs, outputs=x)
+```
+
+Note that the number of filters on the last layer is set to the number of . This will be one output channel per class.output_channels.
+
+### Train the model
+
+Now, all that is left to do is to compile and train the model. Since this is a multiclass classification problem, use the tf.keras.losses.CategoricalCrossentropy loss function with the argument set to , since the labels are scalar integers instead of vectors of scores for each pixel of every class.from_logitsTrue. When running inference, the label assigned to the pixel is the channel with the highest value. This is what the function is doing.create_mask.
```
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:204]: tf.keras.losses.CategoricalCrossentropy should be in a pair of \`

Comment (reveurmichael) [reply open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:204]: for other class/methods from python/tf, format them the same way, i.e. `tf.keras.losses.CategoricalCrossentropy` instead of tf.keras.losses.CategoricalCrossentropy

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md [hunk 4]
```diff
@@ -0,0 +1,973 @@
...
+                                   padding='same',
+                                   kernel_initializer='he_normal')(x)
+
+        outputs = x
+        return models.Model(inputs, outputs, name=self.version)
+```
+
+### SegNet
+
+SegNet is a novel and practical deep fully convolutional neural network architecture for semantic pixel-wise segmentation. This core trainable segmentation engine consists of an encoder network, a corresponding decoder network followed by a pixel-wise classification layer. Here is the structure of SegNet:
```
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:653]: add a ref/citation for SegNet.

Same for other models (e.g. U-Net)

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md (hunk 1 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:50]: Done. It was put at the wrong place.

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md (hunk 2 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:32]: Done.

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md (hunk 3 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:204]: Done.

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md (hunk 4 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:649]: I am not sure what I need to do since SegNet was added to the reference.bib part. Should I add it to the acknowledge part?

Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:661]: Done.

File: open-machine-learning-jupyter-book/deep-learning/image-classification.md [hunk 5]
```diff
@@ -19,6 +19,11 @@ kernelspec:
 
 In this chapter we will introduce the image classification problem, which is the task of assigning an input image one label from a fixed set of categories. This is one of the core problems in Computer Vision that, despite its simplicity, has a large variety of practical applications.
 
+<p style="text-align: center;">
+<iframe src="../assets/html/cls-demo/index.html" width="105%" height="700px;" style="border:none;"></iframe>
+A demo of image classification. <a href="http://vision.stanford.edu/teaching/cs231n/">[source]</a>
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/image-classification.md:24]: why we use source instead of direct bib referring?

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md [hunk 6]
```diff
@@ -0,0 +1,985 @@
...
+---
+
+# Image segmentation
+
+## What is image segmentation?
+
+In an image segmentation task (specifically semantic segmentation), the network assigns a label (or class) to each input image. However, suppose you want to know the shape of that object, which pixel belongs to which object, etc. In this case, you need to assign a class to each pixel of the image—this task is known as segmentation. A segmentation model returns much more detailed information about the image. Image segmentation has many applications in medical imaging, self-driving cars and satellite imaging, just to name a few.
+
+## How to train a model for image segmentation?
+This tutorial uses the Oxford-IIIT Pet Dataset. The dataset consists of images of 37 pet breeds, with 200 images per breed (~100 each in the training and test splits). Each image includes the corresponding labels, and pixel-wise masks. The masks are class-labels for each pixel. Each pixel is given one of three categories:
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:23]: add an empty line

**Review by zhangqi444 — APPROVED** [review]

File: open-machine-learning-jupyter-book/deep-learning/image-segmentation.md (hunk 6 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/image-segmentation.md:24]: Done.

**Review by zhangqi444 — APPROVED** [review]

**Review by reveurmichael — APPROVED** [review]
OK, we will merge first, before proceeding with remaining issues.
